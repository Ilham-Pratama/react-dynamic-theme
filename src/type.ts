export interface Theme<T extends string, V> {
  name: T;
  styles: V;
}

export interface UseTheme<T, V> {
  setCurrentTheme: (arg: T | ((thisArg: T) => T)) => void;
  theme: V;
}

export interface DefaultThemeStyle {
  primary?: string;
  secondary?: string;
  paper?: string;
  background?: string;
  text?: string;
  border?: string;
  notification?: string;
  "spacing-1"?: string | number;
  "spacing-2"?: string | number;
  "spacing-3"?: string | number;
  "spacing-4"?: string | number;
}

export interface CreateThemeArgs<T extends string, V> {
  /**
   * Initial theme name
   *
   * @param
   */
  initialTheme: T;

  /**
   * Theme styles list
   *
   * @param
   */
  themes: Record<T, V>;
}

export interface ThemeProviderArgs {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  children: any;
}

export interface CreateTheme<T> {
  /**
   * Wrap the component that you want to apply the theme to.
   * Make sure to do it first before using the themes.
   *
   * @instance
   */
  ThemeProvider: ({ children }: ThemeProviderArgs) => JSX.Element;

  /**
   * Theme consumer that works best to quickly apply your theme styles.
   *
   * @instance
   */
  ThemeConsumer: React.Consumer<T>;

  /**
   * Custom hook to read and update your theme
   *
   * @instance
   */
  useTheme: () => T;
}
