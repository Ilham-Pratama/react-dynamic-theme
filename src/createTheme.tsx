import React from "react";
import {
  Theme,
  UseTheme,
  CreateThemeArgs,
  ThemeProviderArgs,
  CreateTheme
} from "./type";

const createTheme = <T extends string, V>({
  initialTheme,
  themes
}: CreateThemeArgs<T, V>): CreateTheme<UseTheme<T, Theme<T, V>>> => {
  const ThemeContext = React.createContext<UseTheme<T, Theme<T, V>>>(
    {} as UseTheme<T, Theme<T, V>>
  );

  const ThemeProvider = ({ children }: ThemeProviderArgs) => {
    const [currentTheme, setCurrentTheme] = React.useState<T>(initialTheme);

    const getCurrentTheme = React.useCallback(() => {
      return {
        name: currentTheme,
        styles: themes[currentTheme]
      };
    }, [currentTheme]);

    return (
      <ThemeContext.Provider
        value={{
          setCurrentTheme,
          theme: getCurrentTheme()
        }}
      >
        {children}
      </ThemeContext.Provider>
    );
  };

  const useTheme = () => React.useContext(ThemeContext);

  return { ThemeProvider, ThemeConsumer: ThemeContext.Consumer, useTheme };
};

export default createTheme;
