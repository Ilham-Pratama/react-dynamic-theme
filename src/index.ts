import createTheme from "./createTheme";
import { DefaultThemeStyle } from "./type";

export { createTheme, DefaultThemeStyle };
