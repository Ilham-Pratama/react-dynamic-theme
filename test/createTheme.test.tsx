import "@testing-library/jest-dom";
import React from "react";
import { fireEvent, render, RenderResult } from "@testing-library/react";
import { createTheme } from "../src";
import { DefaultThemeStyle } from "../src/type";

describe("createTheme", () => {
  const { ThemeProvider, ThemeConsumer, useTheme } = createTheme<
    "normal" | "dark",
    DefaultThemeStyle
  >({
    initialTheme: "normal",
    themes: {
      normal: {
        background: "white",
        text: "black"
      },
      dark: {
        background: "black",
        text: "white"
      }
    }
  });

  const AppWithUseTheme = () => {
    const { setCurrentTheme, theme } = useTheme();

    const changeTheme = () =>
      setCurrentTheme((current) => (current === "normal" ? "dark" : "normal"));

    return (
      <div>
        <button
          type="button"
          role="button"
          onClick={changeTheme}
        >
          Change Theme
        </button>
        <p data-testid="theme">{theme.name}</p>
        <p data-testid="background">{theme.styles.background}</p>
        <p data-testid="text">{theme.styles.text}</p>
      </div>
    );
  };

  const AppWithThemeConsumer = () => {
    return (
      <ThemeConsumer>
        {({ setCurrentTheme, theme }) => (
          <>
            <button
              type="button"
              role="button"
              onClick={() => setCurrentTheme("dark")}
            >
              Change Theme
            </button>
            <p data-testid="theme">{theme.name}</p>
            <p data-testid="background">{theme.styles.background}</p>
            <p data-testid="text">{theme.styles.text}</p>
          </>
        )}
      </ThemeConsumer>
    );
  };

  const assertComponent = (
    component: RenderResult<
      typeof import("c:/Open-source/react-dynamic-theme/node_modules/@testing-library/dom/types/queries"),
      HTMLElement
    >
  ) => {
    expect(component).toMatchSnapshot();

    expect(component.getByTestId("theme")).toHaveTextContent("normal");
    expect(component.getByTestId("background")).toHaveTextContent("white");
    expect(component.getByTestId("text")).toHaveTextContent("black");

    fireEvent.click(component.getByRole("button"));

    expect(component.getByTestId("theme")).toHaveTextContent("dark");
    expect(component.getByTestId("background")).toHaveTextContent("black");
    expect(component.getByTestId("text")).toHaveTextContent("white");
  };

  it("renders themed component and updates the theme with useTheme", () => {
    const component = render(
      <ThemeProvider>
        <AppWithUseTheme />
      </ThemeProvider>
    );

    assertComponent(component);
  });

  it("renders themed component and updates the theme with ThemeConsumer", () => {
    const component = render(
      <ThemeProvider>
        <AppWithThemeConsumer />
      </ThemeProvider>
    );

    assertComponent(component);
  });
});
