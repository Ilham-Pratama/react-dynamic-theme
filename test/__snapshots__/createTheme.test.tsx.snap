// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTheme renders themed component and updates the theme with ThemeConsumer 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        role="button"
        type="button"
      >
        Change Theme
      </button>
      <p
        data-testid="theme"
      >
        normal
      </p>
      <p
        data-testid="background"
      >
        white
      </p>
      <p
        data-testid="text"
      >
        black
      </p>
    </div>
  </body>,
  "container": <div>
    <button
      role="button"
      type="button"
    >
      Change Theme
    </button>
    <p
      data-testid="theme"
    >
      normal
    </p>
    <p
      data-testid="background"
    >
      white
    </p>
    <p
      data-testid="text"
    >
      black
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`createTheme renders themed component and updates the theme with useTheme 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <button
          role="button"
          type="button"
        >
          Change Theme
        </button>
        <p
          data-testid="theme"
        >
          normal
        </p>
        <p
          data-testid="background"
        >
          white
        </p>
        <p
          data-testid="text"
        >
          black
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <button
        role="button"
        type="button"
      >
        Change Theme
      </button>
      <p
        data-testid="theme"
      >
        normal
      </p>
      <p
        data-testid="background"
      >
        white
      </p>
      <p
        data-testid="text"
      >
        black
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
